### PR TITLE
[ADP-3226] Use more `IsRecentEra`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request changes some code to align with the decision "Singleton type RecentEra :: AP".

In some places, I have

* Removed an explicit `RecentEra era` argument in favor of the type class / implicit argument `Write.IsRecentEra era` — unless this would have resulted in an ambiguous type, such as for `mkTransaction`.
* Removed `@era` applications that are not of the form `recentEra @era`.
    (I have left applications of the form `shelleyBasedEra @era` and `cardanoEra @era` for convenience, as they are similar in spirit.)

### Comments

* This pull request is an alternative to https://github.com/cardano-foundation/cardano-wallet/pull/4409 .
* I have kept or even added an explicit `RecentEra era` argument to the functions `buildTransaction`, `constructTransaction`, and `constructUnbalancedSharedTransaction`, because their usage site (`Server` module) is prone to ambiguity. The `era` type is introduced through an existential like this

    ```hs
            (Write.PParamsInAnyRecentEra era pp, _)
                <- liftIO $ W.readNodeTipStateForTxWrite netLayer
    ```
  
  In order to get the `era` type into scope, we would need to write

    ```hs
            (Write.PParamsInAnyRecentEra (_era :: RecentEra era) pp, _)
                <- liftIO $ W.readNodeTipStateForTxWrite netLayer
    ```
  
    In both cases, the value or type are needed to specify in which era the `SealedTx` should be created.
    

### Issue Number

ADP-3226

### CODE REVIEW

File: lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 1]
```diff
@@ -1119,7 +1119,8 @@ spec_estimateSignedTxSize = describe "estimateSignedTxSize" $ do
         -> Tx era
         -> IO ()
     test _name bs tx = do
-        let pparams = mockPParamsForBalancing @era
+        let pparams :: Write.PParams era
```
Comment (Anviking) [inline lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:1123]:
```suggestion
        let pparams :: PParams era
```

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 2]
```diff
@@ -1079,29 +1061,27 @@ mkWithdrawals networkId wdrl = case wdrl of
...
+    :: forall era. Write.IsRecentEra era
+    => Cardano.TxBody (CardanoApiEra era)
     -> (XPrv, Passphrase "encryption")
     -> Cardano.KeyWitness (CardanoApiEra era)
-mkShelleyWitness era body key =
+mkShelleyWitness body key =
     Cardano.makeShelleyKeyWitness shelleyEra body (unencrypt key)
   where
-    shelleyEra = Write.shelleyBasedEraFromRecentEra era
+    shelleyEra = Write.shelleyBasedEraFromRecentEra $ Write.recentEra @era
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs]:
```suggestion
    shelleyEra = Write.shelleyBasedEra @era
```

File: lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 3]
```diff
@@ -1181,7 +1182,7 @@ spec_estimateSignedTxSize = describe "estimateSignedTxSize" $ do
         -> Tx era
         -> UTxO era
     utxoPromisingInputsHaveAddress addr tx =
-        utxoFromTxOutsInRecentEra (recentEra @era) $
+        utxoFromTxOutsInRecentEra $
```
Comment (jonathanknowles) [inline lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs]:
```suggestion
        utxoFromTxOutsInRecentEra
```

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 4]
```diff
@@ -1043,19 +1034,10 @@ dummyInput :: TxIn
...
-removeDummyInput RecentEraBabbage = \case
-    Cardano.ShelleyTxBody era body scripts scriptData aux val ->
-        Cardano.ShelleyTxBody
-            era
-            (over inputsTxBodyL (Set.delete (toLedger dummyInput)) body)
-            scripts
-            scriptData
-            aux
-            val
-removeDummyInput RecentEraConway = \case
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:1058]: nice 👍🏻

**Review by jonathanknowles — APPROVED** [review]
This looks good to me!

(Modulo one small comment about a `$` that can be removed.)

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 5]
```diff
@@ -1198,19 +1199,21 @@ prop_sealedTxRecentEraRoundtrip
     sealedTxB = sealedTxFromBytes' currentEra txBytes
 
 makeShelleyTx
-    :: IsRecentEra era
+    :: Write.IsRecentEra era
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:1202]: Why use `IsRecentEra` qualified but not `RecentEra`?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 6]
```diff
@@ -1146,13 +1146,14 @@ binaryCalculationsSpec' era = describe ("calculateBinary - "+||era||+"") $ do
     calculateBinary net utxo outs chgs pairs =
         hex (Cardano.serialiseToCBOR ledgerTx)
       where
+          ledgerTx :: Cardano.Tx (CardanoApiEra era)
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:1149]: `Cannot satisfy: cardano-ledger-binary-1.2.1.0:Cardano.Ledger.Binary.Version.MinVersion <= Ledger.ProtVerHigh era0` without type annotation 🤔 maybe we should have kept the `era` arg here then?

**Review by Anviking — APPROVED** [review]

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 6 above)
Comment (Anviking) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:1149]: This was the error I saw when testing dropping `era` from `constructTransaction` also 🤔

Comment (HeinrichApfelmus) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:1149]: In the presence of GADTs, it is good practice — and sometimes necessary — to annotate polymorphic types in `where` or `let` bindings.

More information can be found in the documentation for the [`MonoLocalBinds`](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/let_generalisation.html) extensions, which is implied by the `GADTs` extension.

In other words: Using `GADTs` has a cost. `MonoLocalBinds` is that cost. We pay it by sometimes adding type signatures.

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 5 above)
Comment (HeinrichApfelmus) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:1202]: Historical reasons only. `RecentEra` was unqualified before I touched the file; I at least wanted to qualify all `IsRecentEra` that I touched.

File: lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs:1123]: It turns out that `PParams` is not in scope. Other functions in the module use `Write.PParams` as well — I'm going to go with the latter.

**Pull request merged**